### PR TITLE
feat: update statistic during query execution

### LINKED ISSUE #9631: Feature: update statistics during query execution
**Summary**

Description for this feature.

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. Update statistics for filter and join during query execution, this will further improve the accuracy of CBO
2. fix right join bug
3. process some flaky tests

Closes #9631

### CODE REVIEW

File: src/query/sql/src/planner/plans/filter.rs [hunk 1]
```diff
@@ -89,12 +90,25 @@ impl Operator for Filter {
             // Compute selectivity for each conjunction
             selectivity *= sb.compute_selectivity(pred);
         }
+        dbg!(input_prop.cardinality);
```
Comment (leiysky) [inline src/query/sql/src/planner/plans/filter.rs:93]: Remove this.

File: src/query/sql/src/planner/optimizer/property/column_stat.rs [hunk 2]
```diff
@@ -38,3 +38,10 @@ pub struct ColumnStat {
     /// Histogram of column
     pub histogram: Option<Histogram>,
 }
+
+#[derive(Debug, Clone)]
+pub struct NewStatistic {
+    pub min: Option<f64>,
+    pub max: Option<f64>,
```
Comment (leiysky) [inline src/query/sql/src/planner/optimizer/property/column_stat.rs:45]: Why are the `min`, `max` f64?

Comment (xudong963) [reply src/query/sql/src/planner/optimizer/property/column_stat.rs:45]: Good catch, use `Datum` is better

Comment (xudong963) [reply src/query/sql/src/planner/optimizer/property/column_stat.rs]: fixed

**Comment by xudong963** [comment]
Strange.., a tpch test occurs wrong answer

```
13_0009_q9:                                                             [ FAIL ] - result differs with:
--- /Users/xudong/owner/databend/tests/suites/0_stateless/13_tpch/13_0009_q9.result	2022-10-18 15:31:02.000000000 +0800
+++ /Users/xudong/owner/databend/tests/suites/0_stateless/13_tpch/13_0009_q9.stdout	2023-01-17 13:55:21.000000000 +0800
@@ -1,5 +1,5 @@
 ALGERIA	1998	2321785.368
 ALGERIA	1997	3685016.858
-ALGERIA	1996	4276597.425
+ALGERIA	1996	4269554.726
 ALGERIA	1995	4418370.415
-ALGERIA	1994	3864849.952
+ALGERIA	1994	3855080.378
```

File: tests/sqllogictests/suites/mode/standalone/explain/join_reorder/star_multiple_tables.test [hunk 3]
```diff
@@ -40,6 +40,8 @@ analyze table t4
 statement ok
 analyze table t5
 
+# A flaky test, analyze table may get different results.
```
Comment (xudong963) [inline tests/sqllogictests/suites/mode/standalone/explain/join_reorder/star_multiple_tables.test:43]: After https://github.com/datafuselabs/databend/pull/9640, we can retry the test @leiysky

**Comment by xudong963** [comment]
> Strange.., a tpch test occurs wrong answer
> 
> ```
> 13_0009_q9:                                                             [ FAIL ] - result differs with:
> --- /Users/xudong/owner/databend/tests/suites/0_stateless/13_tpch/13_0009_q9.result	2022-10-18 15:31:02.000000000 +0800
> +++ /Users/xudong/owner/databend/tests/suites/0_stateless/13_tpch/13_0009_q9.stdout	2023-01-17 13:55:21.000000000 +0800
> @@ -1,5 +1,5 @@
>  ALGERIA	1998	2321785.368
>  ALGERIA	1997	3685016.858
> -ALGERIA	1996	4276597.425
> +ALGERIA	1996	4269554.726
>  ALGERIA	1995	4418370.415
> -ALGERIA	1994	3864849.952
> +ALGERIA	1994	3855080.378
> ```

Precision problem

File: src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_state_impl.rs [hunk 4]
```diff
@@ -305,7 +305,8 @@ impl HashJoinState for JoinHashTable {
         }
 
         if input_block.is_empty() {
-            return Ok(vec![]);
+            let null_block = self.null_blocks_for_right_join(&unmatched_build_indexes)?;
```
Comment (xudong963) [inline src/query/service/src/pipelines/processors/transforms/hash_join/hash_join_state_impl.rs:308]: Accidentally found a right join bug 🤣

**Review by leiysky — APPROVED** [review]

**Pull request merged**